homeassistant/components/zha/__init__.py

```diff
@@ -221,44 +221,78 @@ def device_removed(self, device):
+                await self._attempt_single_cluster_device(
+                    endpoint,
+                    cluster,
+                    profile_clusters[0],
+                    device_key,
+                    zha_const.SINGLE_INPUT_CLUSTER_DEVICE_CLASS,
+                    'in_clusters',
+                    discovered_info,
+                    join,
+                )
 
-                await discovery.async_load_platform(
-                    self._hass,
-                    component,
-                    DOMAIN,
-                    {'discovery_key': cluster_key},
-                    self._config,
+            for cluster in endpoint.out_clusters.values():
+                await self._attempt_single_cluster_device(
+                    endpoint,
+                    cluster,
+                    profile_clusters[1],
+                    device_key,
+                    zha_const.SINGLE_OUTPUT_CLUSTER_DEVICE_CLASS,
+                    'out_clusters',
+                    discovered_info,
+                    join,
                 )
 
     def register_entity(self, ieee, entity_obj):
         """Record the creation of a hass entity associated with ieee."""
         self._device_registry[ieee].append(entity_obj)
 
+    async def _attempt_single_cluster_device(self, endpoint, cluster,
+                                             profile_clusters, device_key,
+                                             device_classes, discovery_attr,
+                                             entity_info, is_new_join):
+        """Try to set up an entity from a "bare" cluster."""
+        if cluster.cluster_id in profile_clusters:
+            return
+        # pylint: disable=unidiomatic-typecheck
+        if type(cluster) not in device_classes:
+            return
+
+        component = device_classes[type(cluster)]
+        cluster_key = "{}-{}".format(device_key, cluster.cluster_id)
+        discovery_info = {
+            'application_listener': self,
+            'endpoint': endpoint,
+            'in_clusters': {},
+            'out_clusters': {},
+            'new_join': is_new_join,
+            'unique_id': cluster_key,
+            'entity_suffix': '_{}'.format(cluster.cluster_id),
+        }
+        discovery_info[discovery_attr] = {cluster.cluster_id: cluster}
+        discovery_info.update(entity_info)
+        self._hass.data[DISCOVERY_KEY][cluster_key] = discovery_info
+
+        await discovery.async_load_platform(
+            self._hass,
+            component,
+            DOMAIN,
+            {'discovery_key': cluster_key},
+            self._config,
+        )
+
 
 class Entity(entity.Entity):
     """A base class for ZHA entities."""
 
     _domain = None  # Must be overridden by subclasses
+    # Normally the entity itself is the listener. Base classes may set this to
+    # a dict of cluster ID -> listener to receive messages for specific
+    # clusters separately
+    _in_listeners = {}
```

Comment:
Having mutable objects as class attributes is dangerous and prone to bugs.

Reply:
Looks like these are used as a class wide registry of listeners. Will there never be a problem of one instance overwriting a listener of another instance? Are cluster_id s unique to each entity?

Reply:
So it looks like we replace the class attribute with an instance attribute of the same name when instantiating the child entity. So then it's not a class wide registry anymore, which is good. But I'd suggest to rewrite this logic anyway. See above about a safer way to handle state update callbacks.

Reply:
It's not ever replaced.

```
~/src/home-assistant/homeassistant/components$ egrep -r '(in|out)_listeners' .
./zha/__init__.py:    _in_listeners = {}
./zha/__init__.py:    _out_listeners = {}
./zha/__init__.py:            cluster.add_listener(self._in_listeners.get(cluster_id, self))
./zha/__init__.py:            cluster.add_listener(self._out_listeners.get(cluster_id, self))
./binary_sensor/zha.py:        self._out_listeners = {
```

Anyways, doing in in async_added_to_hass would be better.

Reply:
The last line in the grep replaces the class attribute with an instance attribute in the entity name space.

Reply:
Uh, I see what you mean. Yes, you're right (except it doesn't need to be an instance attribute).